Try turning on MFEM CUDA tests

NamjaeChoi opened the pull request:
Refs #30915.

NamjaeChoi committed:
Try running MFEM tests on GPU #30915

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3050957/), step Docs: sync website on 3464abe wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31064/site)

This comment will be updated on new commits.

lindsayad committed:
Sync solution vector to host before SetTestVariablesFromTrueVectors

This is what is done for the solution increment and works. Sadly,
I cannot say exactly why it works at this point

lindsayad committed:
Revert "Sync solution vector to host before SetTestVariablesFromTrueVectors"

This reverts commit 5e71368c0616698bb8c9ddc0c1e9a49d10e2550a.

lindsayad committed:
A better fix that doesn't rely on alternating device-host syncs

With this fix the following things happen:
- We sync the memory flags of the solution vector to the grid function
  truth vector which has a data alias to the solution vector
- With the properly set memory flags of the true vector, we do a true
  device to device transfer between the solution vector and the grid
  function... which because these are just aliases of each other this
  is a no-op. Previously the true vector memory flags were not sync'd
  and the grid function thought the vector was on-host so it did the
  memory transfer from there ... and the host memory just had a 0
  initial condition
- We longer need an explicit HostRead in MFEMTransient because the data
  collection handles that automatically for us if we need up-to-date data
  on the host

alexanderianblair commented on framework/src/mfem/problem_operators/ProblemOperatorInterface.C:
```diff
@@ -48,7 +51,13 @@ ProblemOperatorInterface::SetTestVariablesFromTrueVectors()
+    // We must sync the memory flags from the true true vector to the grid function copy of the true
+    // vector
+    mooseAssert(_test_true_vector, "The true vector should already have been set");
+    test_var->GetTrueVector().SyncMemory(*_test_true_vector);
```
I'm not sure I understand this - `X` is the vector storing all (true) grid function DoFs. If `_test_variables.size()` is greater than 1, `_test_true_vector` will be longer than the last `test_var` true DoF vector

lindsayad replied:
It is a little silly/rendudant to do this for all `_test_variables`. We should just do this for the zeroth test variable

lindsayad replied:
I'm not sure what the `Registered` flag is for in MFEM, but I do know that `flags & Registered` is `true` so we never do the check in the `MFEM_ASSERT`

alexanderianblair replied:
Understood - this makes sense now with the usage of `_test_true_vector` formed in `Init`. I agree each grid function needs its own flags synced with X.

lindsayad committed:
Sync block vector from actual grid function

The true vector at this point in the simulation is a dummy vector

lindsayad committed:
More flag fixes

- Make sure that we sync the global block vector flags to the sub-block
  vector flags
- Also, immediately after setting the true vector to our X vector,
  sync the flags and call SetFromTrueVector

NamjaeChoi commented:
Looks like our CUDA machine is too busy.

lindsayad committed:
Conduit data output needs explicit host read

lindsayad committed:
Don't run hypre solver tests in parallel

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3066228/), step Generate coverage on 0587b76 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e981aa</th>
      <th colspan=4>#31064 0587b7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.52%</td>
      <td>85.52%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>115771</td>
      <td>115785</td>
      <td>+14</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19604</td>
      <td>19602</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31064/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31064/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31064/coverage/xfem)

</details>



This comment will be updated on new commits.

NamjaeChoi commented:
Tests passed!

nmnobre commented on test/tests/mfem/kernels/tests:
```diff
@@ -199,6 +210,8 @@
     capabilities = 'mfem'
+    devices = 'cpu cuda'
     recover = false
+    max_parallel = 1 # mfem-hypre doesn't appear to work with multiple ranks with cuda
```
Let's rebase so we don't miss any tests. Given pretty much all tests require `max_parallel = 1`, I wonder if we should move this comment up to the beginning of the file or somewhere else.

lindsayad replied:
We have a bit of interweaving issues. The `XMLDiffs` innately can't be used with more than one rank based on conversations with @alexanderianblair. And then we have this MOOSE-MFEM-HypreGMRES issue

alexanderianblair replied:
> We have a bit of interweaving issues. The `XMLDiffs` innately can't be used with more than one rank based on conversations with @alexanderianblair. And then we have this MOOSE-MFEM-HypreGMRES issue

This is true, since the partitioning may differ between runs, and the number of output files in a `ParaViewDataCollection` will depend on the number of processors used.

We have a couple of `VectorPostProcessors` inbound for sampling MFEM FE variables at arbitrary points using  GSLib from @k-collie - which should enable us to convert many of the existing XMLDiff tests over to CSVDiff tests (also significantly reducing output sizes...) - WIP branch is [here](https://github.com/aurora-multiphysics/platypus-moose/tree/k-collie/mfem-point-and-line-sampler)

nmnobre reviewed (changes requested):
Haven't reviewed/tested the new syncing changes/logic yet.

lindsayad committed:
Explicitly annotate why MFEMCurlCurl/Gravity can't use cuda

lindsayad committed:
Add devices for new MFEMDarcy test

lindsayad committed:
Have to assign blocks of X from grid function true vector

Otherwise the code doesn't work in parallel. This makes sure the assignment
of the blocks of X are properly distributed in parallel

lindsayad committed:
Annotate tests that schemadiff

lindsayad committed:
Remaining serial restriction annotations

moosebuild commented:
Job [CUDA GPU](https://civet.inl.gov/job/3068719/) on 87bf9d5 : invalidated by @lindsayad

lindsayad committed:
Don't use libmesh_ignore in MFEMDataCollection

lindsayad committed:
Don't duplicate operations or code in ProblemOperatorInterface

moosebuild commented:
Job [Apptainer moose-openmpi](https://civet.inl.gov/job/3070239/) on 25735d8 : invalidated by @lindsayad

lindsayad commented:
@nmnobre can you re-review when you have a chance?

nmnobre committed:
Allow subapps to inherit device settings from parent app

lindsayad reviewed (approved):
I approve the changes that aren't mine

nmnobre committed:
Initialise _test_var_names even for non-transient problems

Loads of redundancy in SetGridFunctions() methods across static
and transient class versions, but at least it's verbatim redundancy
now, which I find less confusing.

nmnobre commented:
Likewise, I approve the changes that aren't mine.
